Split binary to booleans

danopz opened the pull request:
Since no one wrote on #191 I tried myself to map the status to boolean types. I already changed the code to use the new boolean fields instead of the status.
I don't really know what to do with StatusFlag::typeToStatus - if we should completely remove this one and somehow fix the code belongs to this, or move this to later and destruct this status in ItemMapper (see added comment). This would also change methods signatures for many methods.
I didn't find a way how we could add a data migraion from status to the new boolean fields, I just found asome [docs in OC](https://doc.owncloud.org/server/10.0/developer_manual/app/schema.html) but it looks like in [NC we don't have this feature](https://docs.nextcloud.com/server/12.0/developer_manual/app/schema.html) - so how should we migrate?

I'm submitting this for the questions and also for a first review.

Todo
---
- [x] Drop unit tests ,write integration tests
- [x] Migration/Repair script
- [x] Completely remove status or map on some code parts?
- [x] Fix the integer/boolean stuff in the queries

sonologic commented:
I was already working on it (assigned it to myself), got about as far as you did :) if you want to work on it further, i can look at some of the other bugs (if Bernhard hasn't fixed them all before that :).

Bernhard pointed me to the docs for migration, its 'repair-steps' in info.cml iirc. At work now, don't have this info at hand.

On 11 July 2017 00:29:58 CEST, Daniel Opitz <notifications@github.com> wrote:
>Since no one wrote on #191 I tried myself to map the status to boolean
>types. I already changed the code to use the new boolean fields instead
>of the status.
>I don't really know what to do with StatusFlag::typeToStatus - if we
>should completely remove this one and somehow fix the code belongs to
>this, or move this to later and destruct this status in ItemMapper (see
>added comment). This would also change methods signatures for many
>methods.
>I didn't find a way how we could add a data migraion from status to the
>new boolean fields, I just found asome [docs in
>OC](https://doc.owncloud.org/server/10.0/developer_manual/app/schema.html)
>but it looks like in [NC we don't have this
>feature](https://docs.nextcloud.com/server/12.0/developer_manual/app/schema.html)
>- so how should we migrate?
>
>I'm submitting this for the questions and also for a first review.
>
>Todo
>---
>- [ ] Fix tests
>- [ ] Migration script
>- [ ] Completely remove status or map on some code parts?
>You can view, comment on, or merge this pull request online at:
>
>  https://github.com/nextcloud/news/pull/203
>
>-- Commit Summary --
>
>  * replaced old status with 2 flags for unread and starred
>
>-- File Changes --
>
>    M lib/Db/FeedMapper.php (12)
>    M lib/Db/Item.php (53)
>    M lib/Db/ItemMapper.php (49)
>    M lib/Db/Mysql/ItemMapper.php (21)
>    M lib/Fetcher/FeedFetcher.php (2)
>    M lib/Service/ItemService.php (9)
>
>-- Patch Links --
>
>https://github.com/nextcloud/news/pull/203.patch
>https://github.com/nextcloud/news/pull/203.diff
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/nextcloud/news/pull/203

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

sonologic commented:
11:38 <@BernhardPosselt> gmc: regarding the binary flag PR
11:38 <@BernhardPosselt> the statusflag class can be killed
11:38 <@BernhardPosselt> all binary flags should be moved to bools in the sql queries
11:39 <@BernhardPosselt> as for migrating: just add 2 new columns and keep the old one
11:39 <@BernhardPosselt> we dont have a way to add migrations yet
11:39 <@BernhardPosselt> will apparently be featured in nextcloud 13
11:39 <@BernhardPosselt> https://github.com/nextcloud/news/blob/master/lib/Db/StatusFlag.php#L30
11:40 <@BernhardPosselt> what the method actually does is that it takes a type and showall flag
11:40 <@BernhardPosselt> the type is a type of feed
11:40 <@BernhardPosselt> e.g. FEED, FOLDER, etc
11:41 <@BernhardPosselt> in case of showAll you want to show all articles
11:41 <@BernhardPosselt> but that filter is not added for starred items
11:41 <@BernhardPosselt> if you display the starred items "feed" you will see read and unread items regardless of your setting
11:42 <@BernhardPosselt> at work right now btw
11:46 <@BernhardPosselt> you can probably replace it with
11:47 <@BernhardPosselt> $showUnread = $showUnread || $feedType === FeedType::STARRED
11:47 <@BernhardPosselt> ehm
11:47 <@BernhardPosselt> $showRead = $showRead || $feedType === FeedType::STARRED

sonologic commented:
Looks like we basically did the same things, except I used true/false in sql where you used 1/0. I tend to prefer true/false.

Anyway, if you want to continue to work on it that would be very welcome!

I started to investigate 'repair-steps' (see http://nextcloudappstore.readthedocs.io/en/latest/developer.html#info-xml), which seems the way to go re migration.

As for the typeToStatus, I figure we should remove it and refactor code that is currently using it. It feels the typeToStatus method is a bit of a kludge..

Thanks for working on this! Let me know if you decide to drop it afterall. I will pause my work on this one for now in favour of some other bug(s).

danopz commented:
Oh sry didn't see you assigned yourself.
Shure I could try to get this done, but also if you want you can still do it yourself, maybe you are more into the codebase.

Indeed True/False are the better option for this.

I will keep working on this until someone stops me from doing this :)

BernhardPosselt commented:
BTW since you are changing SQL statements it might make sense to throw away the unit tests and replace them with integration tests (examples available in the test folder)

danopz commented:
I dropped the unit tests which belongs to the changed mappers and added integration tests.
Changed booleans in sql back to integers (0,1) as sqlite does not support true/false. But now it looks like postgres doesn't allow a boolean = integer check.. so we have to use prepared statements, am I right?

Also I added the repair step which checks for previous installed version. Currently I increased the version by a patch level, I think an increase in minor/major would be better?

danopz commented:
For me it's ready for review.
I don't know why this is failing for MySQL with the message `SQLSTATE[42S22]: Column not found: 1054 Unknown column 'items.unread' in 'where clause'" at`? NC should have added the new fields?

BernhardPosselt commented on lib/Db/Item.php:
```diff
@@ -67,53 +65,54 @@ class Item extends Entity implements IAPI, \JsonSerializable {
-        $this->status |= StatusFlag::UNREAD;
+    public function setUnread($value = true) {
+        $this->markFieldUpdated('unread');
+        $this->unread = (bool)$value;
```
Any reason why we need bool casts?

BernhardPosselt commented:
> I don't know why this is failing for MySQL with the message SQLSTATE[42S22]: Column not found: 1054 Unknown column 'items.unread' in 'where clause'" at? NC should have added the new fields?

I'd take a look at the MySql version on travis. Does it work for you locally on mysql?

BernhardPosselt commented:
> I dropped the unit tests which belongs to the changed mappers and added integration tests.
Changed booleans in sql back to integers (0,1) as sqlite does not support true/false. But now it looks like postgres doesn't allow a boolean = integer check.. so we have to use prepared statements, am I right?

Yep. I'm unsure if the doctrince DAL handles the conversion for you or it could be solved via query builder. Worst case scenario is that we need to create a method that turns bools into whatever format your db currently expects. Check https://github.com/nextcloud/news/blob/master/lib/Db/MapperFactory.php#L23 for how to find out the db

danopz commented:
- Removed the bool casts, they they were there for no real reason
- Shame on me for the MySQL error, this was just a copy-paste mistake
- For the SQL queries this was already fixed by using the booleans as params, DBAL handles this for us

BernhardPosselt commented:
Can you take care of the tests?

danopz commented:
Didn't notice the tests fail again, maybe this was the reason for bool casts. I'll have a look.

danopz commented:
It's green :grin: 
I added empty checks to Item::is* methods which shouldn't be to expensive.

BernhardPosselt commented:
Cool, just a quick question: Can you set up a news install without the pull request, add some feeds and star them, then switch to your branch and ensure that everything still works as expected? Just trying to make sure that everything still works since the PR is quite big.

BernhardPosselt commented:
From a quick glance over the code the only thing that I don't really like is the usage of empty(). That function is pretty tricky and can lead to lots of bugs, see http://php.net/manual/en/function.empty.php

danopz commented:
> From a quick glance over the code the only thing that I don't really like is the usage of empty(). That "function" is pretty tricky and can lead to lots of bugs, see http://php.net/manual/en/function.empty.php

Do you have another idea how to have another way to force booleans? I have in mind that empty checks are faster than bool casts. And without such cast the one unit test was failing. Maybe the problem comes from another place but don't know right now.

danopz commented:
There still is an issue in the migration where status 6 results in unread=false. I didn't find a proper solution yet..

BernhardPosselt commented:
> Do you have another idea how to have another way to force booleans? I have in mind that empty checks are faster than bool casts. And without such cast the one unit test was failing. Maybe the problem comes from another place but don't know right now.

PHP is slow anyways so don't worry about speed of function calls  but speed of algorithmic complexity ;)

There still is an issue in the migration where status 6 results in unread=false. I didn't find a proper solution yet..

> There still is an issue in the migration where status 6 results in unread=false. I didn't find a proper solution yet..

Are we sure that all three dbs treat this as truthy?

    (status & 4)

I'd expect something like WHERE (status & 4) = 4

danopz commented:
> Cool, just a quick question: Can you set up a news install without the pull request, add some feeds and star them, then switch to your branch and ensure that everything still works as expected? Just trying to make sure that everything still works since the PR is quite big.

Tested it on MySQL and SQLite - it was working for me.

> I'd expect something like...

I dropped usage of the QueryBuilder in favor of a native SQL query in the migration step. Now I'm also using `(status & 4) = 4`.

> PHP is slow anyways so don't worry about speed of function calls but speed of algorithmic complexity ;) Also better to investigate the source of the issue

I will have a deeper look.

danopz commented:
@BernhardPosselt I changed something in my last commit to remove bool casts and empty checks.

BernhardPosselt commented:
Ok, reviewed it more closely and lets test it out, as in: merge this and check the effects on our own instances :D

BernhardPosselt commented:
Seems to have worked fine on postgres, well done :) let's give this more testing before we create a release
